server/api.go

```diff
@@ -0,0 +1,399 @@
+	ctx := context.Background()
+	conf := p.getOAuthConfig()
+
+	code := r.URL.Query().Get("code")
+	if len(code) == 0 {
+		http.Error(w, "missing authorization code", http.StatusBadRequest)
+		return
+	}
+
+	state := r.URL.Query().Get("state")
+
+	if storedState, err := p.API.KVGet(state); err != nil {
+		http.Error(w, "missing stored state", http.StatusBadRequest)
+		return
+	} else if string(storedState) != state {
+		http.Error(w, "invalid state", http.StatusBadRequest)
+		return
+	}
+
+	userID := strings.Split(state, "_")[1]
+
+	p.API.KVDelete(state)
+
+	tok, err := conf.Exchange(ctx, code)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	githubClient := p.githubConnect(*tok)
+	gitUser, _, err := githubClient.Users.Get(ctx, "")
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	if user, err := p.API.GetUser(userID); err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	} else {
+		if user.Props == nil {
+			user.Props = model.StringMap{}
+		}
+		user.Props["git_user"] = *gitUser.Login
+		_, err = p.API.UpdateUser(user)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+	}
+
+	userInfo := &GitHubUserInfo{
+		UserID:         userID,
+		Token:          tok,
+		GitHubUsername: gitUser.GetLogin(),
+		LastToDoPostAt: model.GetMillis(),
+		Settings: &UserSettings{
+			SidebarButtons: SETTING_BUTTONS_TEAM,
+			DailyReminder:  true,
+			Notifications:  true,
+		},
+	}
+
+	if err := p.storeGitHubUserInfo(userInfo); err != nil {
+		mlog.Error(err.Error())
+		http.Error(w, "Unable to connect user to GitHub", http.StatusInternalServerError)
+		return
+	}
+
+	if err := p.storeGitHubToUserIDMapping(gitUser.GetLogin(), userID); err != nil {
+		mlog.Error(err.Error())
+	}
+
+	// Post intro post
+	message := fmt.Sprintf("#### Welcome to the Mattermost GitHub Plugin!\n You've connected your Mattermost account to [%s](%s) on GitHub. Read about the features of this plugin below:\n\n##### Daily Reminders\nThe first time you log in each day, you will get a post right here letting you know what messages you need to read and what pull requests are awaiting your review.\nTurn off reminders with `/github settings reminders off`.\n\n##### Notifications\nWhen someone mentions you, requests your review, comments on or modifies one of your pull requests/issues, or assigns you, you'll get a post here about it.\nTurn off notifications with `/github settings notifications off`.\n\n##### Sidebar Buttons\nCheck out the buttons in the left-hand sidebar of Mattermost.\n* The first button tells you how many pull requests are awaiting your review\n* The second tracks the number of unread messages you have\n* The third will refresh the numbers\n\nClick on them!\n\n##### Slash Commands\n"+strings.Replace(COMMAND_HELP, "|", "`", -1), gitUser.GetLogin(), gitUser.GetHTMLURL())
```

Comment:
nit: backticks would be easier to parse here

Reply:
Would be but you can't escape backticks inside backticks unfortunately (hence my whole replacing | with backticks for COMMAND_HELP)

Reply:
Aha, learned something new! 👍